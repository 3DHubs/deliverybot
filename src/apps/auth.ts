import Octokit from "@octokit/rest";
import { Response, Request, NextFunction } from "express";
import request from "request";
import { Application } from "express";
import { CLIENT_ID, CLIENT_SECRET, BASE_URL } from "../config";

export interface AuthedRequest extends Request {
  user?: {
    id: string;
    username: string;
    token: string;
    github: Octokit;
    repo?: { owner: string; repo: string; id: string };
  };
}

export function setUser(req: AuthedRequest, res: Response, next: NextFunction) {
  const token = req.session && req.session.token;
  if (!token) {
    return next();
  }
  const username = req.session!.login;
  const id = req.session!.id;
  if (!id || !username) {
    return next();
  }
  const github = new Octokit({ auth: token });
  req.user = { id, github, token, username };
  next();
}

export function authenticate(
  req: AuthedRequest,
  res: Response,
  next: NextFunction
) {
  const token = req.session && req.session.token;
  if (!token) {
    (req as any).log.info("unauthenticated access redirecting");
    res.redirect("/login");
    return;
  }
  setUser(req, res, next);
}

export async function canWrite(
  gh: Octokit,
  { owner, repo, username }: { owner: string; repo: string; username: string }
): Promise<boolean> {
  const perms = await gh.repos.getCollaboratorPermissionLevel({
    owner,
    repo,
    username
  });
  return ["admin", "write"].includes(perms.data.permission);
}

export async function verifyRepo(
  req: AuthedRequest,
  res: Response,
  next: NextFunction
) {
  const { owner, repo } = req.params;
  const username = req.user!.username;
  const octokit = req.user!.github;
  if (
    req.session &&
    req.session.verified &&
    req.session.verified[`${owner}/${repo}`]
  ) {
    const id = req.session.verified[`${owner}/${repo}`];
    req.user!.repo = { owner, repo, id };
    next();
    return;
  }

  const repoData = await octokit.repos.get({ owner, repo });
  const id = `${repoData.data.id}`;
  const write = await canWrite(octokit, { owner, repo, username });
  if (!write) {
    (req as any).log.info({ owner, repo }, "no write access");
    res.redirect("/login");
    return;
  }
  req.session!.verified = req.session.verified || {};
  req.session.verified[`${owner}/${repo}`] = id;
  req.user!.repo = { owner, repo, id };
  next();
}

function accessToken(code: string): Promise<string> {
  return new Promise((resolve, reject) => {
    request.post(
      {
        headers: {
          accept: "application/json"
        },
        url: "https://github.com/login/oauth/access_token",
        form: {
          client_id: CLIENT_ID,
          client_secret: CLIENT_SECRET,
          code
        }
      },
      (err: Error, _: request.Response, body: any) => {
        if (err) {
          return reject(err);
        }

        const data = JSON.parse(body);
        const token = data["access_token"];
        resolve(token);
      }
    );
  });
}

export async function callback(req: Request, res: Response) {
  const token = await accessToken(req.query.code);
  const octokit = new Octokit({ auth: token });
  const user = await octokit.users.getAuthenticated({});

  req.session!.token = token;
  req.session!.id = user.data.id;
  req.session!.login = user.data.login;
  res.redirect("/");
}

const loginUrl = `https://github.com/login/oauth/authorize?client_id=${CLIENT_ID}&redirect_uri=${BASE_URL}/login/cb`;

export function auth(app: Application) {
  app.get("/login", (req: Request, res: Response) => res.redirect(loginUrl));
  app.get("/login/cb", callback);
}
