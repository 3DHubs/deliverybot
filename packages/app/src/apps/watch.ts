import { Response } from "express";
import { AuthedRequest, authenticate, verifyRepo } from "./auth";
import { Dependencies, MessageService, logger } from "@deliverybot/core";

export interface RepoUpdatedMessage {
  id: number;
  updatedAt: string;
}

export function publishRepoUpdated(
  messageService: MessageService,
  repo: { id: number },
) {
  const messageBus = messageService<RepoUpdatedMessage>();
  messageBus
    .publish(`repos/${repo.id}`, {
      id: repo.id,
      updatedAt: new Date().toISOString(),
    })
    .catch(err => {
      logger.error({ error: err }, "failed to publish repo update");
    });
}

/**
 * Watch sets up the message bus app.
 */
export function watch({ app, robot, messageService }: Dependencies) {
  const messageBus = messageService<RepoUpdatedMessage>();

  app.get(
    "/_/watch/:owner/:repo",
    authenticate,
    verifyRepo,
    async (req: AuthedRequest, res: Response) => {
      const user = req.user!;
      const repo = user.repo!;
      req.log.info({ repoId: repo.id, userId: user.id }, "subscribing");
      const token = await messageBus.subscribe(`repos/${repo.id}`, user);
      try {
        await messageBus.poll(`repos/${repo.id}`, user);
        res.json({ continue: true, id: repo.id, token });
      } catch (error) {
        res.json({ continue: false, id: repo.id, token });
      }
    },
  );

  robot.on("*", async context => {
    const repo = context.payload.repository;
    if (!repo) {
      return;
    }
    publishRepoUpdated(messageService, repo);
  });
}
