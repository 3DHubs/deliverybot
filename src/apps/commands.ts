import { Application, Context, Octokit, Logger } from "probot";
import { render } from "../util";
import yaml from "js-yaml";
import { validate } from "jsonschema";
import {
  ReposListDeploymentsResponseItem,
  ReposGetDeploymentResponse,
  PullsGetResponse
} from "@octokit/rest";
import schema from "../schema.json";
import { canWrite } from "./auth";

const previewAnt = "application/vnd.github.ant-man-preview+json";
const previewFlash = "application/vnd.github.flash-preview+json";

function withPreview<T>(arg: T): T {
  (arg as any).headers = { accept: `${previewAnt},${previewFlash}` };
  return arg as T;
}

function logCtx(context: Context, params: any) {
  return {
    context: {
      installation: context.payload.installation,
      repo: context.payload.repository ? context.repo() : undefined
    },
    ...params
  };
}

interface Deployment {
  task: string;
  payload: any;
  environment: string;
  description: string;
  auto_merge: boolean;
}

export interface Target {
  name: string;
  auto_deploy_on: string;

  // Required contexts  are required to be matched across all deployments in the
  // target set. This is so that one deployment does not succeed before another
  // causing the set to fail.
  required_contexts: string[];

  // Environment information must be copied into all deployments.
  transient_environment: boolean;
  production_environment: boolean;

  // Deployments are the list of deployments to trigger.
  deployments: Deployment[];
}

export type Targets = { [k: string]: Target | undefined };

export async function config(
  github: Octokit,
  {
    owner,
    repo,
    ref
  }: {
    owner: string;
    repo: string;
    ref: string;
  }
): Promise<Targets> {
  const content = await github.repos.getContents({
    owner,
    repo,
    ref,
    path: `.github/deploy.yml`
  });
  const conf =
    yaml.safeLoad(Buffer.from(content.data.content, "base64").toString()) || {};
  const validation = validate(conf, schema, {
    propertyName: "config",
    allowUnknownAttributes: true
  });
  if (validation.errors.length > 0) {
    const err = validation.errors[0];
    throw new Error(`${err.property} ${err.message}`);
  }
  for (const key in conf) {
    conf[key].name = key;
    conf[key].deployments = conf[key].deployments || [];
  }
  return conf;
}

function getDeployBody(
  target: Target,
  deployment: Deployment,
  data: any
): Deployment {
  return withPreview({
    task: deployment.task || "deploy",
    transient_environment: target.transient_environment || false,
    production_environment: target.production_environment || false,
    environment: render(deployment.environment || "production", data),
    auto_merge: deployment.auto_merge || false,
    required_contexts: target.required_contexts || [],
    description: render(deployment.description, data),
    payload: {
      target: target.name,
      ...render(deployment.payload, data)
    }
  });
}

async function handlePRDeploy(context: Context, command: string) {
  context.log.info(logCtx(context, { command }), "pr deploy: handling command");
  try {
    const target = command.split(" ")[1];
    const pr = await context.github.pulls.get({
      owner: context.payload.repository.owner.login,
      repo: context.payload.repository.name,
      pull_number: context.payload.issue.number
    });

    const write = await canWrite(
      context.github,
      context.repo({ username: context.payload.comment.user.login })
    );
    if (!write) {
      context.log.info(logCtx(context, {}), "pr deploy: no write priviledges");
      return;
    }

    await deployCommit(
      context.github,
      context.log,
      context.repo({
        target,
        ref: pr.data.head.ref,
        sha: pr.data.head.sha,
        pr: pr.data
      })
    );
  } catch (error) {
    await context.github.issues.createComment({
      owner: context.payload.repository.owner.login,
      repo: context.payload.repository.name,
      issue_number: context.payload.issue.number,
      body: `:rotating_light: Failed to trigger deployment. :rotating_light:\n${error.message}`
    });
  }
}

/**
 * Deploy commit handles all the necessities of creating a conformant deployment
 * including templating and more. All deploys should go through this function.
 * We need to deploy always using the ref of a branch so that finding
 * deployments later we can query using the branch ref.
 */
export async function deployCommit(
  github: Octokit,
  log: Logger,
  {
    owner,
    repo,
    target,
    ref,
    sha,
    pr
  }: {
    owner: string;
    repo: string;
    target: string;
    ref: string;
    sha: string;
    pr?: PullsGetResponse;
  }
) {
  const logCtx = {
    deploy: { target, ref, pr },
    context: { repo: { owner, repo } }
  };
  const commit = await github.git.getCommit({ owner, repo, commit_sha: sha });

  // Params are the payload that goes into every deployment - change these in a
  // backwards compatible way always.
  const params = {
    ref,
    target,
    owner,
    repo,
    short_sha: sha.substr(0, 7),
    commit: commit.data,
    pr: pr ? pr.number : undefined,
    pull_request: pr
  };

  const conf = await config(github, { owner, repo, ref });
  const targetVal = conf[target];
  if (!targetVal) {
    log.info(logCtx, "deploy: failed - no target");
    throw new Error(`Deployment target "${target}" does not exist`);
  }
  if (targetVal.deployments.length === 0) {
    log.info(logCtx, "deploy: failed - no deployments");
    throw new Error(`Deployment target "${target}" has no deployments`);
  }

  const deployed = [];
  for (const deployment of targetVal.deployments) {
    const body = {
      owner,
      repo,
      ref,
      ...getDeployBody(targetVal, deployment, params)
    };
    try {
      log.info({ ...logCtx, body }, "deploy: deploying");
      // TODO: Handle auto_merge case correctly here.
      // https://developer.github.com/v3/repos/deployments/#merged-branch-response
      const deploy = await github.repos.createDeployment(body);
      log.info({ ...logCtx, body }, "deploy: successful");
      deployed.push(deploy.data as ReposGetDeploymentResponse);
    } catch (error) {
      log.error({ ...logCtx, error, body }, "deploy: failed");
      throw error;
    }
  }
  return deployed;
}

async function handleAutoDeploy(context: Context) {
  context.log.info("auto deploy: checking deployments");
  const config = await context.config("deploy.yml");
  for (const key in config) {
    const deployment = config[key];
    await autoDeployTarget(context, key, deployment);
  }
}

async function autoDeployTarget(
  context: Context,
  target: string,
  targetVal: Target
) {
  const autoDeploy = targetVal.auto_deploy_on;
  if (!autoDeploy) {
    return;
  }
  const ref = autoDeploy.replace("refs/", "");
  context.log.info(logCtx(context, { ref }), "auto deploy: verifying");
  const refData = await context.github.git.getRef(context.repo({ ref }));
  const sha = refData.data.object.sha;

  const deploys = await context.github.repos.listDeployments(
    context.repo({ sha })
  );
  const environments = targetVal.deployments.map(d => d.environment);
  if (deploys.data.find(d => environments.includes(d.environment))) {
    context.log.info(logCtx(context, { ref }), "auto deploy: already deployed");
    return;
  }

  context.log.info(logCtx(context, { ref }), "auto deploy: deploying");
  try {
    await deployCommit(
      context.github,
      context.log,
      context.repo({
        ref,
        sha,
        target
      })
    );
    context.log.info(logCtx(context, { ref }), "auto deploy: done");
  } catch (error) {
    context.log.error(
      context.repo({ error, ref, target }),
      "auto deploy: failed"
    );
  }
}

async function handlePRClose(context: Context) {
  const ref = context.payload.pull_request.head.ref;
  const sha = context.payload.pull_request.head.sha;
  const deployments = await context.github.repos.listDeployments(
    withPreview({ ...context.repo(), ref })
  );
  context.log.info(logCtx(context, { ref }), "pr close: listed deploys");

  // List all deployments for this pull request by environment to undeploy the
  // last deployment for every environment.
  const environments: { [env: string]: ReposListDeploymentsResponseItem } = {};
  for (const deployment of deployments.data.reverse()) {
    // Only terminate transient environments.
    if (!deployment.transient_environment) {
      context.log.info(
        logCtx(context, { ref, deployment: deployment.id }),
        "pr close: not transient"
      );
      continue;
    }
    try {
      context.log.info(
        logCtx(context, { ref, deployment: deployment.id }),
        "pr close: mark inactive"
      );
      await context.github.repos.createDeploymentStatus(
        withPreview({
          ...context.repo(),
          deployment_id: deployment.id,
          state: "inactive"
        })
      );
    } catch (error) {
      context.log.error(
        logCtx(context, { error, ref, deployment: deployment.id }),
        "pr close: marking inactive failed"
      );
    }
    environments[deployment.environment] = deployment;
  }

  context.log.info(
    logCtx(context, { ref, environments: Object.keys(environments).map(e => e) }),
    "pr close: remove deploys"
  );
  for (const env in environments) {
    const deployment = environments[env];
    try {
      context.log.info(
        logCtx(context, { ref, deployment: deployment.id }),
        "pr close: remove deploy"
      );
      // Undeploy for every unique environment by copying the deployment params
      // and triggering a deployment with the task "remove".
      await context.github.repos.createDeployment(
        context.repo({
          ref: sha,
          task: "remove",
          required_contexts: [],
          payload: deployment.payload as any,
          environment: deployment.environment,
          description: deployment.description || "",
          transient_environment: deployment.transient_environment,
          production_environment: deployment.production_environment
        })
      );
    } catch (error) {
      context.log.error(
        logCtx(context, { error, ref, deployment: deployment.id }),
        "pr close: failed to undeploy"
      );
    }
  }
}

export function commands(app: Application) {
  app.on("push", async context => {
    await handleAutoDeploy(context);
  });
  app.on("status", async context => {
    await handleAutoDeploy(context);
  });
  app.on("check_run", async context => {
    await handleAutoDeploy(context);
  });
  app.on("issue_comment.created", async context => {
    if (context.payload.comment.body.startsWith("/deploy")) {
      await handlePRDeploy(context, context.payload.comment.body);
    }
  });
  app.on("pull_request.closed", async context => {
    await handlePRClose(context);
  });
}
