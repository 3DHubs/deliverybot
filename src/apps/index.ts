import { auth } from "./auth";
import { deploy } from "./deploy";
import { home } from "./home";
import { commands } from "./commands";

export const apps = [home, auth, deploy, commands];
