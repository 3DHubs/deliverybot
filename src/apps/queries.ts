import { request as ghRequest } from "@octokit/request";
import { Targets } from "./commands";

async function gql(token: string, query: string, variables: any) {
  const resp = await ghRequest({
    data: { query, variables },
    method: "POST",
    url: "/graphql",
    headers: {
      authorization: `token ${token}`,
      accept: "application/vnd.github.antiope-preview+json"
    }
  });
  if (resp.data.errors && resp.data.errors.length > 0) {
    throw new Error(
      `Graphql request failure: ${JSON.stringify(resp.data.errors)}`
    );
  }
  return resp.data.data;
}

const CommitData = `
messageHeadline oid message
author { user { login } }
status { contexts { state } }
checkSuites(last: 50) { nodes { conclusion status } }
deployments(last: 50) {
  nodes {
    id environment description
    creator { login }
    latestStatus { logUrl state }
  }
}
`;

const CommitQuery = `
query commit(
  $owner: String!, $repo: String!, $oid: GitObjectID
) {
  repository(owner:$owner,name:$repo) {
    node: object(oid: $oid) {
      ... on Commit { ${CommitData} }
    }
  }
}
`;

const CommitsQuery = `
query commits(
  $owner: String!, $repo: String!, $branch: String!
) {
  repository(owner:$owner,name:$repo) {
    refs(refPrefix:"refs/heads/",first:50) { nodes { name } }
    ref(qualifiedName:$branch) {
      target {
        ... on Commit { history(first: 10) { edges { node { ${CommitData} }}}}
      }
    }
  }
}
`;

export async function commits(
  token: string,
  owner: string,
  repo: string,
  target: string,
  branch: string
) {
  const result = await gql(token, CommitsQuery, { owner, repo, branch });
  return View(owner, repo, target, branch, result);
}

export async function commit(
  token: string,
  owner: string,
  repo: string,
  target: string,
  branch: string,
  oid: string
) {
  const result = await gql(token, CommitQuery, { owner, repo, oid });
  return Commit(owner, repo, target, branch, result.repository);
}

export function View(
  owner: string,
  repo: string,
  target: string,
  branch: string,
  data: any
) {
  const branches = Branches(branch, data.repository.refs.nodes);
  const commits = data.repository.ref.target.history.edges.map((edge: any) =>
    Commit(owner, repo, target, branch, edge)
  );
  return { commits, branches };
}

const failing = ["FAILURE", "FAILING", "ERROR", "CANCELLED", "TIMED_OUT"];
const pending = ["PENDING", "ACTION_REQUIRED", "QUEUED", "IN_PROGRESS"];
const success = ["SUCCESS"];
const waiting = ["WAITING"];

export function AggregateStatus(statuses: string[]) {
  if (statuses.find(s => failing.includes(s.toUpperCase()))) {
    return "FAILURE";
  }
  if (statuses.find(s => pending.includes(s.toUpperCase()))) {
    return "PENDING";
  }
  if (statuses.find(s => success.includes(s.toUpperCase()))) {
    return "SUCCESS";
  }
  if (statuses.find(s => waiting.includes(s.toUpperCase()))) {
    return "WAITING";
  }
  return "WAITING";
}

export function Check(node: any) {
  const statuses: string[] = ((node.status && node.status.contexts) || [])
    .map((ctx: any) => ctx.state)
    .concat(
      node.checkSuites.nodes.map((cs: any) => cs.conclusion || cs.status)
    );
  return { status: Status(AggregateStatus(statuses)) };
}

export function Status(status?: string) {
  const s = [(status || "WAITING").toUpperCase()];
  switch (AggregateStatus(s)) {
    case "FAILURE":
      return { failure: true, color: "red" };
    case "PENDING":
      return { pending: true, color: "yellow" };
    case "SUCCESS":
      return { success: true, color: "green" };
    default:
      return { waiting: true, color: "gray" };
  }
}

export function Deployment(node: any) {
  const statuses = (node.deployments || []).nodes.map(
    (deploy: any) =>
      (deploy.latestStatus && deploy.latestStatus.state) || "WAITING"
  );
  return { status: Status(AggregateStatus(statuses)) };
}

export function Undeployed(node: any) {
  const deployment = Deployment(node);
  const check = Check(node);
  if (check.status.success && deployment.status.waiting) {
    return true;
  }
  return false;
}

export function Deployments(node: any) {
  return (node.deployments || []).nodes.map((deploy: any) => ({
    status: Status(deploy.latestStatus && deploy.latestStatus.state),
    description: truncate(deploy.description, 20),
    environment: deploy.environment,
    creator: deploy.creator.login,
    url: deploy.latestStatus && deploy.latestStatus.logUrl
  }));
}

function truncate(s: string | undefined | null, t: number) {
  if (!s) {
    return ""
  }
  if (s.length > t) {
    return s.substr(0, t) + "..."
  }
  return s
}

export function Commit(
  owner: string,
  repo: string,
  target: string,
  branch: string,
  edge: any
) {
  return {
    owner,
    repo,
    target,
    branch,
    message: edge.node.messageHeadline,
    oid: edge.node.oid,
    oidShort: edge.node.oid.substr(0, 7),
    author: edge.node.author.user.login,
    undeployed: Undeployed(edge.node),
    deployment: Deployment(edge.node),
    deployments: Deployments(edge.node),
    check: Check(edge.node)
  };
}

export function Branches(
  active: string,
  branches: any[]
): Array<{ name: string; active: boolean }> {
  return branches.map(n => ({
    name: n.name,
    active: active === n.name
  }));
}

export function Targets(target: string, conf: Targets | null) {
  return (
    conf &&
    Object.keys(conf).map((name: string) => ({
      name,
      active: name === target
    }))
  );
}
