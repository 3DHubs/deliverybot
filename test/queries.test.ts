import { View } from "../src/apps/queries";

const DATE_TO_USE = new Date(1568298661772);
const _Date = Date;
// @ts-ignore
global.Date = jest.fn(() => DATE_TO_USE);
global.Date.UTC = _Date.UTC;
global.Date.parse = _Date.parse;
global.Date.now = _Date.now;

describe("Queries", () => {
  it("renders a payload", () => {
    // Patch the following into apps/queries to generate:
    // writeFileSync("test/query.json", JSON.stringify(result));
    const query = require("./fixtures/query.json");
    expect(
      View("colinjfw", "deliverybot-example", "production", "master", query)
    ).toMatchSnapshot();
  });
});
