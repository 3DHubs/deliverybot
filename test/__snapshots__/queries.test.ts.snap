// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries renders a payload 1`] = `
Object {
  "branches": Array [
    Object {
      "active": false,
      "name": "colinjfw-patch-1",
    },
    Object {
      "active": false,
      "name": "colinjfw-patch-2",
    },
    Object {
      "active": false,
      "name": "colinjfw-patch-3",
    },
    Object {
      "active": false,
      "name": "colinjfw-patch-4",
    },
    Object {
      "active": true,
      "name": "master",
    },
  ],
  "commits": Array [
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployments": Array [],
      "message": "Update deploy.yml",
      "oid": "244a4459e999bae6730f12a6203f35df32c41a32",
      "oidShort": "244a445",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "https://console.cloud.google.com/cloud-build/builds/01ecc3eb-1a88-428a-ba0a-d421c90eb360?project=deliverybot",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "f28d26327642e6ec54f9af6a46a2ef18ca6313dd",
      "oidShort": "f28d263",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "https://console.cloud.google.com/cloud-build/builds/b68f76c5-31f0-47e7-a28d-a1bebec9fc39?project=deliverybot",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "03938e013a96875a18b5fdedf4bc98178b01b37b",
      "oidShort": "03938e0",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "yellow",
            "pending": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": null,
        },
      ],
      "message": "Update deploy.yml",
      "oid": "266a09824a55b13d3f0e75313e637aaff12c6468",
      "oidShort": "266a098",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployments": Array [],
      "message": "Update deploy.yml",
      "oid": "13b08bb27476b7062cd2554dcc925816c4e69e0c",
      "oidShort": "13b08bb",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161479032",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "fc245f759c563fc6fe119837d7fb7d1c09f1b944",
      "oidShort": "fc245f7",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440765",
        },
        Object {
          "creator": "colinjfw",
          "environment": "staging",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440865",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "67504c8be5ee318eb1e1514fb5601b6250935839",
      "oidShort": "67504c8",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440458",
        },
        Object {
          "creator": "colinjfw",
          "environment": "staging",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440575",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "fd3d28ad40d3dd5b6f5a553f3c287cec32dd87b7",
      "oidShort": "fd3d28a",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/89cf477a-f7d0-4a84-948e-ee8fa535f5b4",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/32ddb9ab-27cb-415c-94ce-461d843cd7d3",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/de3f0b69-cbd5-4f52-a8de-5d73167619d1",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/317ab83f-80ee-474e-8e09-38887ee79194",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/52b74663-c848-497b-b52b-ca5425b035e3",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/63a8cc8c-ed24-4d79-b97f-0c70ce6a8f9e",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/2e9ec0e5-98d9-4b1c-86e7-2086d3764be8",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161324570",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161324837",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161325022",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161325741",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161337388",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161337565",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161338055",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161338198",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161338704",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161345711",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "eb7930eef65f379b9c6a91f7ff790bcda91b245e",
      "oidShort": "eb7930e",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "branch": "master",
      "check": Object {
        "status": Object {
          "color": "yellow",
          "pending": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "yellow",
          "pending": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "yellow",
            "pending": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/c94697a5-dc47-4c7f-b632-2240ae8b7d34",
        },
      ],
      "message": "Create deploy.yml",
      "oid": "626285ee3f420e0c7a03f36ef01080a60224d13a",
      "oidShort": "626285e",
      "owner": "deliverybot",
      "repo": "example",
      "target": "production",
      "undeployed": false,
    },
  ],
}
`;
